### PR TITLE
add api basic auth support

### PR DESCRIPTION
**Pull request by ninuxer**
Implement basic auth for api interface base on gin middleware.

Although gin has its own gin.BasicAuth() middleware, but I still implemented it by myself for future expansion.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -55,6 +60,7 @@ type Config struct {
 	Zookeeper   *zookeeper.Config `yaml:"zookeeper"`
 	Admin       AdminConfig       `yaml:"admin"`
 	Controller  *ControllerConfig `yaml:"controller"`
+	ApiAuth     *ApiAuthConfig    `yaml:"apiAuth"`
```
Comment (torwig) [inline config/config.go]: I see that we we snake_case for yaml-tags in this file. Could you please adhere this convention?

File: config/config.go [hunk 2]
```diff
@@ -37,6 +37,11 @@ type AdminConfig struct {
 	Addr string `yaml:"addr"`
 }
 
+type ApiAuthConfig struct {
+	ApiUser     string `yaml:"apiUser"`
+	ApiPassword string `yaml:"apiPassword"`
```
Comment (torwig) [inline config/config.go]: The same: yaml-tags.

File: server/route.go [hunk 3]
```diff
@@ -36,6 +36,14 @@ func (srv *Server) initHandlers() {
 		c.Set(consts.ContextKeyStore, srv.store)
 		c.Next()
 	}, middleware.RedirectIfNotLeader)
+
+	if srv.config.ApiAuth.ApiUser != "" && srv.config.ApiAuth.ApiPassword != "" {
```
Comment (torwig) [inline server/route.go]: This condition can be extracted in a single method of config like config.AuthConfigured().

**Comment by torwig** [comment]
How to prove that the implementation is correct without tests?

**Comment by ninuxer** [comment]
I will fix it, thanks

> How to prove that the implementation is correct without tests?

**Comment by ninuxer** [comment]
already fix and commit，my test case see
image

**Comment by ninuxer** [comment]
Close the pr as it did not take into account the authentication of the web UI.

**Pull request closed without merging**